File: backtestApp/panels/profile_panel.py
from PyQt5.QtWidgets import (QGroupBox, QVBoxLayout, QHBoxLayout, QLabel, 
                            QPushButton, QComboBox)
from PyQt5.QtGui import QFont
from .base_panel import BasePanel

# import ConfigManager
from config_manager import ConfigManager

class ProfilePanel(BasePanel):
    """Panel de gestion des profils de configuration."""
    
    def __init__(self, parent=None):
        super().__init__(parent)
        self.config_manager = parent.config_manager if parent else ConfigManager()
    
    def create(self):
        """Crée le panel de gestion des profils."""
        # Créer le groupe pour les profils
        profile_group = QGroupBox("Profils de configuration")
        profile_layout = QVBoxLayout()
        
        # Affichage du profil actif
        if self.config_manager:
            current_profile = self.config_manager.current_profile
        else:
            current_profile = "DEFAULT"
            
        self.widgets['current_profile_label'] = QLabel(f"Profil actif: {current_profile}")
        self.widgets['current_profile_label'].setStyleSheet("font-weight: bold;")
        profile_layout.addWidget(self.widgets['current_profile_label'])
        
        # Actions de profil
        profile_buttons_layout = QHBoxLayout()
        
        # Bouton pour sauvegarder le profil actuel
        save_profile_btn = QPushButton("💾 Sauvegarder")
        save_profile_btn.setToolTip("Sauvegarder les paramètres actuels dans le profil courant")
        if self.parent and self.config_manager:
            save_profile_btn.clicked.connect(
                lambda: self.config_manager.save_current_profile(self.parent, self.parent)
            )
        profile_buttons_layout.addWidget(save_profile_btn)
        
        # Bouton pour créer un nouveau profil
        new_profile_btn = QPushButton("➕ Nouveau")
        new_profile_btn.setToolTip("Créer un nouveau profil avec les paramètres actuels")
        if self.parent and self.config_manager:
            new_profile_btn.clicked.connect(
                lambda: self.config_manager.prompt_create_new_profile(self.parent, self.parent)
            )
        profile_buttons_layout.addWidget(new_profile_btn)
        
        # Bouton pour supprimer un profil
        delete_profile_btn = QPushButton("🗑️ Supprimer")
        delete_profile_btn.setToolTip("Supprimer le profil actuel")
        if self.parent and self.config_manager:
            delete_profile_btn.clicked.connect(
                lambda: self.config_manager.delete_current_profile(self.parent, self.parent)
            )
        profile_buttons_layout.addWidget(delete_profile_btn)
        
        profile_layout.addLayout(profile_buttons_layout)
        
        # Menu déroulant pour sélectionner un profil
        profile_selector_layout = QHBoxLayout()
        profile_selector_layout.addWidget(QLabel("Charger un profil:"))
        
        self.widgets['profile_combo'] = QComboBox()
        if self.config_manager:
            self.widgets['profile_combo'].addItems(self.config_manager.list_profiles())
            self.widgets['profile_combo'].setCurrentText(self.config_manager.current_profile)
            if self.parent:
                self.widgets['profile_combo'].currentTextChanged.connect(self.parent.load_selected_profile)
        profile_selector_layout.addWidget(self.widgets['profile_combo'])
        
        profile_layout.addLayout(profile_selector_layout)
        
        # Ajout des boutons d'import/export de configuration
        import_export_layout = QHBoxLayout()
        
        # Bouton d'importation
        import_btn = QPushButton("📥 Importer")
        import_btn.setToolTip("Importer une configuration depuis un fichier")
        if self.parent and self.config_manager:
            import_btn.clicked.connect(
                lambda: self.config_manager.import_config_from_file(self.parent, self.parent)
            )
        import_export_layout.addWidget(import_btn)
        
        # Bouton d'exportation
        export_btn = QPushButton("📤 Exporter")
        export_btn.setToolTip("Exporter la configuration vers un fichier")
        if self.parent and self.config_manager:
            export_btn.clicked.connect(
                lambda: self.config_manager.export_config_to_file(self.parent)
            )
        import_export_layout.addWidget(export_btn)
        
        # Ajouter ce layout au layout des profils
        profile_layout.addLayout(import_export_layout)
        
        # Finaliser le groupe de profils
        profile_group.setLayout(profile_layout)
        return profile_group
        
    def get_values(self):
        """Récupère les valeurs des widgets du panel."""
        return {
            'profile': self.widgets['profile_combo'].currentText() if 'profile_combo' in self.widgets else ""
        }
        
    def set_values(self, values):
        """Définit les valeurs des widgets du panel."""
        if 'profile' in values and 'profile_combo' in self.widgets:
            index = self.widgets['profile_combo'].findText(values['profile'])
            if index >= 0:
                self.widgets['profile_combo'].setCurrentIndex(index)

File: backtest_app_cpp/src/panels/profile_panel.cpp
#include "profile_panel.h"
#include "../config_manager.h"
#include "../app.h"
#include <QFont>
#include <QDebug>

ProfilePanel::ProfilePanel(QWidget* parent)
    : QObject(parent), BasePanel(parent)
{
    // Récupérer le gestionnaire de configuration depuis le parent App
    if (parent) {
        App* mainApp = qobject_cast<App*>(parent->window());
        if (mainApp) {
            m_configManager = mainApp->getConfigManager();
        } else {
            qWarning() << "Impossible de récupérer l'instance App";
            m_configManager = nullptr;
        }
    } else {
        m_configManager = nullptr;
    }
}

QGroupBox* ProfilePanel::create()
{
    // Créer le groupe pour les profils
    QGroupBox* profileGroup = new QGroupBox("Profils de configuration");
    QVBoxLayout* profileLayout = new QVBoxLayout();
    
    // Affichage du profil actif
    QString currentProfile = "DEFAULT";
    if (m_configManager) {
        currentProfile = m_configManager->getCurrentProfile();
    }
    
    m_widgets["current_profile_label"] = new QLabel(QString("Profil actif: %1").arg(currentProfile));
    static_cast<QLabel*>(m_widgets["current_profile_label"])->setStyleSheet("font-weight: bold;");
    profileLayout->addWidget(static_cast<QLabel*>(m_widgets["current_profile_label"]));
    
    // Actions de profil (boutons)
    QHBoxLayout* profileButtonsLayout = new QHBoxLayout();
    
    // Bouton pour sauvegarder le profil actuel
    QPushButton* saveProfileBtn = new QPushButton("💾 Sauvegarder");
    saveProfileBtn->setToolTip("Sauvegarder les paramètres actuels dans le profil courant");
    
    // CORRECTION : Utiliser une connexion directe avec vérification
    connect(saveProfileBtn, &QPushButton::clicked, this, [this]() {
        if (m_configManager && m_parent) {
            App* mainApp = qobject_cast<App*>(m_parent->window());
            if (mainApp) {
                bool success = m_configManager->saveCurrentProfile(mainApp);
                if (success) {
                    updateProfileUI(m_configManager->getCurrentProfile());
                }
            }
        }
    });
    
    profileButtonsLayout->addWidget(saveProfileBtn);
    
    // Bouton pour créer un nouveau profil
    QPushButton* newProfileBtn = new QPushButton("➕ Nouveau");
    newProfileBtn->setToolTip("Créer un nouveau profil avec les paramètres actuels");
    
    connect(newProfileBtn, &QPushButton::clicked, this, [this]() {
        if (m_configManager && m_parent) {
            App* mainApp = qobject_cast<App*>(m_parent->window());
            if (mainApp) {
                bool success = m_configManager->promptCreateNewProfile(mainApp);
                if (success) {
                    // Mettre à jour la liste des profils
                    QComboBox* combo = static_cast<QComboBox*>(m_widgets["profile_combo"]);
                    if (combo) {
                        combo->clear();
                        combo->addItems(m_configManager->listProfiles());
                        combo->setCurrentText(m_configManager->getCurrentProfile());
                    }
                    updateProfileUI(m_configManager->getCurrentProfile());
                }
            }
        }
    });
    
    profileButtonsLayout->addWidget(newProfileBtn);
    
    // Bouton pour supprimer un profil
    QPushButton* deleteProfileBtn = new QPushButton("🗑️ Supprimer");
    deleteProfileBtn->setToolTip("Supprimer le profil actuel");
    
    connect(deleteProfileBtn, &QPushButton::clicked, this, [this]() {
        if (m_configManager && m_parent) {
            App* mainApp = qobject_cast<App*>(m_parent->window());
            if (mainApp) {
                bool success = m_configManager->deleteCurrentProfile(mainApp);
                if (success) {
                    // Mettre à jour la liste des profils
                    QComboBox* combo = static_cast<QComboBox*>(m_widgets["profile_combo"]);
                    if (combo) {
                        combo->clear();
                        combo->addItems(m_configManager->listProfiles());
                        combo->setCurrentText(m_configManager->getCurrentProfile());
                    }
                    updateProfileUI(m_configManager->getCurrentProfile());
                }
            }
        }
    });
    
    profileButtonsLayout->addWidget(deleteProfileBtn);
    
    profileLayout->addLayout(profileButtonsLayout);
    
    // Menu déroulant pour sélectionner un profil
    QHBoxLayout* profileSelectorLayout = new QHBoxLayout();
    profileSelectorLayout->addWidget(new QLabel("Charger un profil:"));
    
    m_widgets["profile_combo"] = new QComboBox();
    if (m_configManager) {
        QComboBox* combo = static_cast<QComboBox*>(m_widgets["profile_combo"]);
        combo->addItems(m_configManager->listProfiles());
        combo->setCurrentText(m_configManager->getCurrentProfile());
        
        // Connecter le signal de changement
        connect(combo, QOverload<const QString&>::of(&QComboBox::currentTextChanged),
                this, &ProfilePanel::loadSelectedProfile);
    }
    profileSelectorLayout->addWidget(static_cast<QComboBox*>(m_widgets["profile_combo"]));
    
    profileLayout->addLayout(profileSelectorLayout);
    
    // Boutons d'import/export
    QHBoxLayout* importExportLayout = new QHBoxLayout();
    
    // Bouton d'importation
    QPushButton* importBtn = new QPushButton("📥 Importer");
    importBtn->setToolTip("Importer une configuration depuis un fichier");
    connect(importBtn, &QPushButton::clicked, this, [this]() {
        if (m_configManager && m_parent) {
            App* mainApp = qobject_cast<App*>(m_parent->window());
            if (mainApp) {
                m_configManager->importConfigFromFile(mainApp);
            }
        }
    });
    importExportLayout->addWidget(importBtn);
    
    // Bouton d'exportation
    QPushButton* exportBtn = new QPushButton("📤 Exporter");
    exportBtn->setToolTip("Exporter la configuration vers un fichier");
    connect(exportBtn, &QPushButton::clicked, this, [this]() {
        if (m_configManager && m_parent) {
            App* mainApp = qobject_cast<App*>(m_parent->window());
            if (mainApp) {
                m_configManager->exportConfigToFile(mainApp);
            }
        }
    });
    importExportLayout->addWidget(exportBtn);
    
    profileLayout->addLayout(importExportLayout);
    
    // Finaliser le groupe de profils
    profileGroup->setLayout(profileLayout);
    return profileGroup;
}

QMap<QString, QVariant> ProfilePanel::getValues()
{
    QMap<QString, QVariant> values;
    // Ce panel ne stocke pas de valeurs directement
    return values;
}

void ProfilePanel::setValues(const QMap<QString, QVariant>& values)
{
    Q_UNUSED(values);
    // Ce panel ne définit pas de valeurs directement
}

void ProfilePanel::loadSelectedProfile(const QString& profileName)
{
    if (m_configManager && profileName != m_configManager->getCurrentProfile()) {
        App* mainApp = qobject_cast<App*>(m_parent->window());
        if (mainApp) {
            bool success = m_configManager->applyProfileToUI(profileName);
            if (success) {
                updateProfileUI(profileName);
            } else {
                // En cas d'échec, rétablir le profil précédent dans le combobox
                QComboBox* combo = static_cast<QComboBox*>(m_widgets["profile_combo"]);
                if (combo) {
                    combo->blockSignals(true);
                    combo->setCurrentText(m_configManager->getCurrentProfile());
                    combo->blockSignals(false);
                }
            }
        }
    }
}

void ProfilePanel::updateProfileUI(const QString& profileName)
{
    // Mettre à jour l'étiquette du profil actif
    QLabel* label = static_cast<QLabel*>(m_widgets["current_profile_label"]);
    if (label) {
        label->setText(QString("Profil actif: %1").arg(profileName));
    }
    
    // Mettre à jour la sélection dans le combo
    QComboBox* combo = static_cast<QComboBox*>(m_widgets["profile_combo"]);
    if (combo) {
        combo->blockSignals(true);
        combo->setCurrentText(profileName);
        combo->blockSignals(false);
    }
}